### ISSUE TITLE
Less readable when a promise .then takes certain arguments

### ISSUE THREAD

**Issue by benjaminapetersen** [prettier/prettier#682]
Just found about prettier this morning, looks like a great package!  Playing with it and overall pretty pleased with the way things look.  I did come across this issue that seems to be a bit less readable with the output.

Original 

```JavaScript
SomeService
  .get('things')
  .then(_.spread((firstThing, secondThing) => {
    // do stuff with things
  }));
```

Formatted output

```JavaScript
SomeService.get("things").then(
  _.spread((firstThing, secondThing) => {
    // do stuff with things
  })
);
```

Its harder to see in the formatted that `_.spread` is doing the job of deconstructing a single argument into multiple arguments given to the callback to the `.then` fn.  It seems something that you could get used to, but is perhaps less ideal.

**Comment by vjeux** [comment]
Thanks for the feedback!

Right now we special case arrow functions to be on the same line but not `_.spread` and we are unlikely going to do so.

However, you can use the es6 arguments spread to achieve the same effect which prettier will output as you would like:

```js
SomeService.get("things").then(([firstThing, secondThing]) => {
  // do stuff with things
});
```

Hopefully browser/node compatibility with this construct is good enough that you can use it in your codebase.

**Comment by benjaminapetersen** [comment]
Makes sense!  I was thinking it might be an issue more with `.then(` breaking the line, but when I make the structure more complex I think the formatter handles it well:

Original
```JavaScript
SomeService
  .get('things')
  .then(shizzle(otherFn(stuff((foo, bar) => {
    // do stuff with foo & bar
  }))));

```

Formatted output:
```JavaScript
SomeService.get("things").then(
  shizzle(
    otherFn(
      stuff((foo, bar) => {
        // do stuff with foo & bar
      })
    )
  )
);
```

I work on a project that does not yet use es6 or a transpiler, which is nice, but def like that the es6 spread is an eventual viable solution.

**Comment by kjbekkelund** [comment]
(Not sure if this fits in this issue, but adding it as an example from a `.then` in my code base)

Simplified example from my code base where I feel a `.then` became less readable:

```js
function someFunction(url) { 
  return get(url)
    .then(
      json => dispatch(success(json)),
      error => dispatch(failed(error))
    );
}
```

turns into:

```js
function someFunction(url) {
  return get(url).then(json => dispatch(success(json)), error =>
    dispatch(failed(error)));
}
```

https://jlongster.github.io/prettier/#%7B%22content%22%3A%22function%20someFunction(url)%20%7B%20%5Cn%20%20return%20get(url)%5Cn%20%20%20%20.then(%5Cn%20%20%20%20%20%20json%20%3D%3E%20dispatch(success(json))%2C%5Cn%20%20%20%20%20%20error%20%3D%3E%20dispatch(failed(error))%5Cn%20%20%20%20)%3B%5Cn%7D%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22doc%22%3Afalse%7D%7D

In this case it feels like it should break "outside" (newline before `.then`) before it breaks "inside" (second fn within `then`), but haven't looked into what effects that could have on other things though.

**Comment by vjeux** [comment]
Thanks for the report, I think I know how to fix this. If there are more than one function, then don't do the last argument expansion. I'll try and implement it this week and see how it feels

**Comment by yamafaktory** [comment]
#809 doesn't fix it actually.

**Comment by vjeux** [comment]
We're unlikely going to support treating _.spread specially since es6 spread exists and is properly printed, so I'm going to go ahead and close this issue. @kjbekkelund your issue has been fixed by #809

**Comment by benjaminapetersen** [comment]
Sounds good, thx!
